File: app/src/main/java/se/gustavkarlsson/aurora_notifier/android/background/providers/impl/aggregating_aurora_factors/GetGeomagLocation.java
package se.gustavkarlsson.aurora_notifier.android.background.providers.impl.aggregating_aurora_factors;

import android.location.Location;
import android.util.Log;

import se.gustavkarlsson.aurora_notifier.android.background.providers.GeomagLocationProvider;
import se.gustavkarlsson.aurora_notifier.android.models.factors.GeomagLocation;

class GetGeomagLocation implements DefaultingCallable<GeomagLocation> {
	private static final String TAG = GetGeomagLocation.class.getSimpleName();

	private final GeomagLocationProvider provider;
	private final Location location;

	GetGeomagLocation(GeomagLocationProvider provider, Location location) {
		this.provider = provider;
		this.location = location;
	}

	@Override
	public GeomagLocation call() throws Exception {
		Log.i(TAG, "Getting geomagnetic location...");
		GeomagLocation geomagLocation = provider.getGeomagLocation(location.getLatitude(), location.getLongitude());
		Log.d(TAG, "Geomagnetic location is: " + geomagLocation);
		return geomagLocation;
	}

	@Override
	public GeomagLocation getDefault() {
		return new GeomagLocation();
	}
}

File: app/src/main/kotlin/se/gustavkarlsson/skylight/android/services_impl/providers/aggregating_aurora_factors/GetGeomagLocation.kt
package se.gustavkarlsson.skylight.android.services_impl.providers.aggregating_aurora_factors

import android.location.Location
import org.jetbrains.anko.AnkoLogger
import org.jetbrains.anko.debug
import org.jetbrains.anko.info
import se.gustavkarlsson.skylight.android.services.providers.GeomagLocationProvider
import se.gustavkarlsson.skylight.android.entities.GeomagLocation
import java.util.concurrent.Callable
import java.util.concurrent.TimeoutException

class GetGeomagLocation(
	private val provider: GeomagLocationProvider,
	private val location: Location
) : ErrorHandlingTask<GeomagLocation>, AnkoLogger {

    override val callable: Callable<GeomagLocation>
        get() = Callable { this.call() }

    private fun call(): GeomagLocation {
        info("Getting geomagnetic location...")
        val geomagLocation = provider.getGeomagLocation(location.latitude, location.longitude)
        debug("Geomagnetic location is: $geomagLocation")
        return geomagLocation
    }

    override fun handleInterruptedException(e: InterruptedException) = GeomagLocation()

    override fun handleThrowable(e: Throwable) = GeomagLocation()

    override fun handleTimeoutException(e: TimeoutException) = GeomagLocation()
}
